### PR TITLE
Improve compile performance

### PR DESCRIPTION
**Pull request by joelfisch**
The compile performance on my computer is improved by around 20% (from 7 minutes to 5.6 minutes). With the .hs-boot files, there is no longer a big memory requirement for compilation.

- https://github.com/Haskell-OpenAPI-Code-Generator/Haskell-OpenAPI-Client-Code-Generator/issues/22
- https://github.com/Haskell-OpenAPI-Code-Generator/Haskell-OpenAPI-Client-Code-Generator/issues/2
- https://github.com/Haskell-OpenAPI-Code-Generator/Haskell-OpenAPI-Client-Code-Generator/issues/10

### CODE REVIEW

File: .circleci/golden/expected/src/OpenAPI/Types/Cat.hs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    = CatAnother_relativeCat Cat
+    | CatAnother_relativePetByType PetByType
+    | CatAnother_relativeText Data.Text.Internal.Text
+    deriving (GHC.Show.Show, GHC.Classes.Eq)
+instance Data.Aeson.Types.ToJSON.ToJSON CatAnother_relativeVariants
+    where toJSON (CatAnother_relativeCat a) = Data.Aeson.Types.ToJSON.toJSON a
+          toJSON (CatAnother_relativePetByType a) = Data.Aeson.Types.ToJSON.toJSON a
+          toJSON (CatAnother_relativeText a) = Data.Aeson.Types.ToJSON.toJSON a
+instance Data.Aeson.Types.FromJSON.FromJSON CatAnother_relativeVariants
+    where parseJSON val = case Data.Aeson.Types.FromJSON.fromJSON val of
```
Comment (NorfairKing) [inline .circleci/golden/expected/src/OpenAPI/Types/Cat.hs:68]: Wouldn't using `<|>` be much simpler than multiple levels of these cases?

File: .circleci/golden/expected/src/OpenAPI/Types/Dog.hs [hunk 2]
```diff
@@ -90,20 +91,16 @@ data DogBreed
...
+          toJSON (DogBreedEnumString_Dingo) = "Dingo"
+          toJSON (DogBreedEnumString_Husky) = "Husky"
+          toJSON (DogBreedEnumString_Retriever) = "Retriever"
+          toJSON (DogBreedEnumString_Shepherd) = "Shepherd"
+instance Data.Aeson.Types.FromJSON.FromJSON DogBreed
+    where parseJSON val = GHC.Base.pure (if | val GHC.Classes.== "Dingo" -> DogBreedEnumString_Dingo
+                                            | val GHC.Classes.== "Husky" -> DogBreedEnumString_Husky
+                                            | val GHC.Classes.== "Retriever" -> DogBreedEnumString_Retriever
+                                            | val GHC.Classes.== "Shepherd" -> DogBreedEnumString_Shepherd
+                                            | GHC.Base.otherwise -> DogBreedEnumOther val)
```
Comment (NorfairKing) [inline .circleci/golden/expected/src/OpenAPI/Types/Dog.hs:106]: This is _so_ much nicer :p

File: .circleci/golden/expected/src/OpenAPI/Types/Cat.hs (hunk 1 above)
Comment (joelfisch) [reply .circleci/golden/expected/src/OpenAPI/Types/Cat.hs:68]: Yes, this would be a nice improvement! Thank you for the tip ;-)

**Comment by joelfisch** [comment]
> Super cool. Can you explain how the hs-boot files work?

The hs-boot files allow to break the import cycles as they contain `data` and `type` declarations without a body. GHC then compiles (where necessary, namely where the `{-# SOURCE #-}` pragma appears in imports) the hs-boot to hi-boot and uses this file to compile the hs files in turn (see https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/separate_compilation.html#how-to-compile-mutually-recursive-modules).

In our case it allows the splitting of all (root) models into their own module even if they contain cyclic imports.

**Pull request merged**